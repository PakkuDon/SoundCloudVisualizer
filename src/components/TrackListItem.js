import React from 'react'
import PropTypes from 'prop-types'
import { SoundCloudTrack } from '../propTypes'

import IconDelete from './IconDelete'
import IconPlay from './IconPlay'
import IconQueue from './IconQueue'

const TrackListItem = ({ track }) => (
  <div className="track">
    <div>{track.user.username}</div>
    <div>{track.title}</div>
    <div>{track.genre}</div>
    <div>
      <a href={track.permalink_url}>Permalink</a>
    </div>
    <div className="actions">
      <button><IconPlay /></button>
      <button><IconDelete /></button>
      <button><IconQueue /></button>
    </div>
  </div>
)

TrackListItem.propTypes = {
  track: SoundCloudTrack.isRequired,
}

export default TrackListItem
