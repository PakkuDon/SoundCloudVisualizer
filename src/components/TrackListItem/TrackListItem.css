@import "../../styles/colors";

.root {
  border-bottom: 1px solid $blue300;
  padding: 5px;

  &:hover {
    background-color: $blue300;
  }
}

.actions {
  text-align: right;
}
