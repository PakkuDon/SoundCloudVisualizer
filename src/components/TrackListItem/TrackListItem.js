import React from "react"
import PropTypes from "prop-types"
import { SoundCloudTrack } from "../../propTypes"

import Button from "../Button"
import IconDelete from "../Icons/IconDelete"
import IconPlay from "../Icons/IconPlay"
import IconQueue from "../Icons/IconQueue"

import styles from "./TrackListItem.css"

const TrackListItem = ({ onDelete, onQueue, onSelect, track }) => (
  <div className={styles.root}>
    <div>
      <a
        href={track.user.permalink_url}
        target="_blank"
        rel="noopener noreferrer"
      >
        {track.user.username}
      </a>
    </div>
    <div>
      <a href={track.permalink_url} target="_blank" rel="noopener noreferrer">
        {track.title}
      </a>
    </div>
    <div>{track.genre}</div>
    <div className={styles.actions}>
      {onSelect && (
        <Button onClick={() => onSelect(track.id)}>
          <IconPlay title="Play track" />
        </Button>
      )}
      {onDelete && (
        <Button onClick={() => onDelete(track.id)}>
          <IconDelete title="Delete track" />
        </Button>
      )}
      {onQueue && (
        <Button onClick={() => onQueue(track.id)}>
          <IconQueue title="Add to playback queue" />
        </Button>
      )}
    </div>
  </div>
)

TrackListItem.propTypes = {
  track: SoundCloudTrack.isRequired,
  onDelete: PropTypes.func,
  onQueue: PropTypes.func,
  onSelect: PropTypes.func,
}

export default TrackListItem
