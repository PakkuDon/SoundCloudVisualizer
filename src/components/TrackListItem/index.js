import React from 'react'
import PropTypes from 'prop-types'
import { SoundCloudTrack } from '../../propTypes'

import Button from '../Button'
import IconDelete from '../icons/IconDelete'
import IconPlay from '../icons/IconPlay'
import IconQueue from '../icons/IconQueue'

import styles from './styles.css'

const TrackListItem = ({ track }) => (
  <div className={styles.root}>
    <div>{track.user.username}</div>
    <div>{track.title}</div>
    <div>{track.genre}</div>
    <div>
      <a href={track.permalink_url}>Permalink</a>
    </div>
    <div className={styles.actions}>
      <Button><IconPlay /></Button>
      <Button><IconDelete /></Button>
      <Button><IconQueue /></Button>
    </div>
  </div>
)

TrackListItem.propTypes = {
  track: SoundCloudTrack.isRequired,
}

export default TrackListItem
