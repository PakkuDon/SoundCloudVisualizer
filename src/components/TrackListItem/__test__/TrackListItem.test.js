import React from "react"
import { render, fireEvent } from "@testing-library/react"

import TrackListItem from "../TrackListItem"
import TrackFactory from "../../../testHelpers/trackFactory"

describe("TrackListItem", () => {
  it("renders track information", () => {
    const { container } = render(
      <TrackListItem track={TrackFactory.create()} />,
    )

    expect(container).toMatchSnapshot()
  })

  describe("when onSelect callback is provided", () => {
    it("calls onSelect when play button is clicked", () => {
      const spy = jest.fn()
      const { getByTitle } = render(
        <TrackListItem track={TrackFactory.create()} onSelect={spy} />,
      )

      fireEvent.click(getByTitle("Play track"))

      expect(spy).toBeCalled()
    })
  })

  describe("when onDelete callback is provided", () => {
    it("calls onDelete when delete button is clicked", () => {
      const spy = jest.fn()
      const { getByTitle } = render(
        <TrackListItem track={TrackFactory.create()} onDelete={spy} />,
      )

      fireEvent.click(getByTitle("Delete track"))

      expect(spy).toBeCalled()
    })
  })

  describe("when onQueue callback is provided", () => {
    it("calls onQueue when queue button is clicked", () => {
      const spy = jest.fn()
      const { getByTitle } = render(
        <TrackListItem track={TrackFactory.create()} onQueue={spy} />,
      )

      fireEvent.click(getByTitle("Add to playback queue"))

      expect(spy).toBeCalled()
    })
  })
})
