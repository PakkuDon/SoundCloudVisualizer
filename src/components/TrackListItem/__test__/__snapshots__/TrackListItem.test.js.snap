// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TrackListItem renders track information 1`] = `
<div>
  <div
    class="root"
  >
    <div>
      Bob
    </div>
    <div>
      Hello World
    </div>
    <div />
    <div>
      <a
        href=""
        rel="noopener noreferrer"
        target="_blank"
      >
        Permalink
      </a>
    </div>
    <div
      class="actions"
    />
  </div>
</div>
`;
